rotkehlchen/chain/ethereum/modules/curve/decoder.py

```diff
@@ -231,6 +257,109 @@ def _decode_curve_deposit_events(
+        Then match and label events.
+        Then reshuffle events to make sure that spend and receive are consecutive.
+
+        Note that `sold_token_address` and `bought_token_address` are not always found (e.g.
+        when pool for some reason is not present in our cache). If tokens that were swapped are
+        detected then we use them when iterating over `decoded_events` list and matching transfers.
+        If they are not detected then conditions when matching transfer events are a bit broader.
+        """
+
+        # These are nullable because in case a curve pool is not stored in our cache or if it
+        # is a swap in a metapool (TOKEN_EXCHANGE_UNDERLYING) we will skip token check.
+        sold_token_address: Optional[ChecksumEvmAddress] = None
+        bought_token_address: Optional[ChecksumEvmAddress] = None
+
+        swapping_contract: ChecksumEvmAddress
+        if context.tx_log.topics[0] in (TOKEN_EXCHANGE, TOKEN_EXCHANGE_UNDERLYING):
+            pool_address = context.tx_log.address
+            swapping_contract = pool_address
+            # When a single pool is used, spender and receiver is always the same
+            spender_address = receiver_address = hex_or_bytes_to_address(context.tx_log.topics[1])
+            sold_token_id = hex_or_bytes_to_int(context.tx_log.data[:32])
+            raw_sold_amount = hex_or_bytes_to_int(context.tx_log.data[32:64])
+            bought_token_id = hex_or_bytes_to_int(context.tx_log.data[64:96])
+            raw_bought_amount = hex_or_bytes_to_int(context.tx_log.data[96:128])
+            if (
+                context.tx_log.topics[0] == TOKEN_EXCHANGE and
+                pool_address in self.curve_pools and
+                len(self.curve_pools[pool_address]) > max(sold_token_id, bought_token_id)  # Make sure that tokens of the pool are cached  # noqa: E501
+            ):
+                sold_token_address = self.curve_pools[pool_address][sold_token_id]
+                bought_token_address = self.curve_pools[pool_address][bought_token_id]
+        else:  # EXCHANGE_MULTIPLE
+            swapping_contract = CURVE_SWAP_ROUTER
+            spender_address = hex_or_bytes_to_address(context.tx_log.topics[1])
+            receiver_address = hex_or_bytes_to_address(context.tx_log.topics[2])
+            raw_sold_amount = hex_or_bytes_to_int(context.tx_log.data[-64:-32])
+            raw_bought_amount = hex_or_bytes_to_int(context.tx_log.data[-32:])
+            # Curve swap router logs route (a list of addresses) that was used. Route consists of
+            # 9 elements. Consider X a number of pools that was used. Then the structure can be
+            # described in the following way:
+            # At 0 index: Address of token in
+            # From 1 to X indices: Addresses of pools that were used
+            # At 1+X index: Address of token out
+            # From 2+X to 8 indices: Unused elements (zero addresses)
```

Comment:
The yoda usage here is confusing

```suggestion
            # At X + 1 index: Address of token out
            # From X + 2 to X + 8 indices: Unused elements (zero addresses)
```